src/main/java/org/opensearch/security/resources/ResourceSharingIndexHandler.java

```diff
@@ -0,0 +1,1175 @@
+     * @param requestUserName The user requesting to revoke the resource
+     * @param shareWith       Updated sharing configuration object containing access control settings:
+     *                        {
+     *                        "action-group": {
+     *                        "users": ["user1", "user2"],
+     *                        "roles": ["role1", "role2"],
+     *                        "backend_roles": ["backend_role1"]
+     *                        }
+     *                        }
+     * @param isAdmin         Boolean indicating whether the user requesting to revoke is an admin or not
+     * @param listener        Listener to be notified when the operation completes
+     * @throws RuntimeException if there's an error during the update operation
+     */
+    public void updateResourceSharingInfo(
+        String resourceId,
+        String sourceIdx,
+        String requestUserName,
+        ShareWith shareWith,
+        boolean isAdmin,
+        ActionListener<ResourceSharing> listener
+    ) {
+        XContentBuilder builder;
+        Map<String, Object> shareWithMap;
+        try {
+            builder = XContentFactory.jsonBuilder();
+            shareWith.toXContent(builder, ToXContent.EMPTY_PARAMS);
+            String json = builder.toString();
+            shareWithMap = DefaultObjectMapper.readValue(json, new TypeReference<>() {
+            });
+        } catch (IOException e) {
+            LOGGER.error("Failed to build json content", e);
+            listener.onFailure(new OpenSearchStatusException("Failed to build json content", RestStatus.INTERNAL_SERVER_ERROR));
+            return;
+        }
+
+        StepListener<ResourceSharing> fetchDocListener = new StepListener<>();
+        StepListener<Boolean> updateScriptListener = new StepListener<>();
+        StepListener<ResourceSharing> updatedSharingListener = new StepListener<>();
+
+        // Fetch resource sharing doc
+        fetchResourceSharingDocument(sourceIdx, resourceId, fetchDocListener);
+
+        // build update script
+        fetchDocListener.whenComplete(currentSharingInfo -> {
+            // Check if user can share. At present only the resource creator and admin is allowed to share the resource
+            if (!isAdmin && currentSharingInfo != null && !currentSharingInfo.getCreatedBy().getCreator().equals(requestUserName)) {
+
+                LOGGER.error("User {} is not authorized to share resource {}", requestUserName, resourceId);
+                listener.onFailure(
+                    new OpenSearchStatusException(
+                        "User " + requestUserName + " is not authorized to share resource " + resourceId,
+                        RestStatus.FORBIDDEN
+                    )
+                );
+            }
+
+            Script updateScript = new Script(ScriptType.INLINE, "painless", """
+                if (ctx._source.share_with == null) {
```

Comment:
The painless script here is hard the grok. Mentally, I think it makes sense to get the doc corresponding to the sharing info of the resource being acted upon, update it in memory and save it back.

I think one challenge is the centralized approach of single .resource-sharing index with sharing info of all sharable resources in the cluster.

One thing that could help is instead of having a single index, to have indexes matching 1-to-1 with the resource index.

So for example if we had a Quip Plugin with Sharable documents then there could be

.quip_documents

and the sharing info backed index managed by the security plugin

.quip_documents_sharing_info

For each index we could keep the same id for documents in both indices so that we can make GET requests instead of searches.

Reply:
The painless script is intended to move the two calls (fetch and update) from security to the indexing eco-system. 

Individual indexes, while separating concerns, may make the management little cumbersome. And our current setup allows us to move into that direction if need be.